File: public/src/app/books/book.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, Subject } from 'rxjs';
import { BookInterface } from './common/BookInterface';
import { BaseService } from '../common/base.service';
import { BookRequest } from './add/add-book.component';
import { PaginatorState } from "primeng/paginator";

export interface BookEditData {
  book: BookInterface;
  listIndex: number;
}

export const $bookChangeSource = new Subject<BookEditData>();


@Injectable({providedIn: 'root'})
export class BookService extends BaseService<any> {

  private booksUrl = this.baseUrl + 'books';
  // headers = new HttpHeaders(header);

  constructor(private http: HttpClient) {
    super(http);
  }

  getPagedListByGenre(params: PaginatorState, genreId: number) {
    let body = {
      "page": params.page,
      "rows": params.rows
    }
    return this.http.post<any>(this.booksUrl  + `/genre/${genreId}` , body);
  }

  getPagedList(params: PaginatorState) {

    let body = {
      "page": params.page,
      "rows": params.rows
    }

    return this.http.post<any>(this.booksUrl  + '/paged-list' , body);
  }

  getAll() {
    return this.http.get<BookInterface>(this.booksUrl);
  }

  getMostPopular() {
    return this.http.get<BookInterface[]>(this.booksUrl + '/most-popular');
  }

    getPromotions() {
      const queryParams = new HttpParams();
      queryParams.append('promotion', true);
      return this.http.get<BookInterface[]>(this.booksUrl + '/promotions', {
        params: queryParams
      });
    }


  getCategories() {
    return this.http.get<any[]>(this.baseUrl + 'categories',  {});
  }

    addBook(body: any) {
      return this.http.post(this.booksUrl, body);
    }

    getById(bookId: number): Observable<BookInterface> {
      return this.http.get<BookInterface>(this.booksUrl + '/' + bookId);
    }

    update (id: any, value: BookRequest) {
      return this.http.post<BookInterface>(this.booksUrl + '/' + id, value);
    }

    findBook(request: any) {
      return this.http.post<BookInterface[]>(this.booksUrl + '/search', request);
    }

    deleteBook(bookId: number) {
      return this.http.delete(this.booksUrl + '/' + bookId);
    }
}

File: backend-java/src/main/java/org/npd21tech/services/BookService.java
package org.npd21tech.services;

import java.util.List;

import org.npd21tech.dtos.BookGenreResponse;
import org.npd21tech.dtos.BookRequest;
import org.npd21tech.dtos.BookResponse;
import org.npd21tech.params.BookSearchParams;
import org.npd21tech.params.PagedList;
import org.npd21tech.params.PagedParams;

public interface BookService {

    PagedList<BookResponse> getPagedList(PagedParams params);

    List<BookResponse> getAll(BookSearchParams bookSearchParams);

    List<BookResponse> getAllPromotion();

    List<BookGenreResponse> getAllGenre();

    void loadBooks();

    BookResponse save(BookRequest bookRequest);

    BookResponse getById(Long id);

    BookResponse update(Long id, BookRequest bookRequest);

    void delete(Long id);

    PagedList<BookResponse> getPagedListByGenre(PagedParams params, Long genreId);
}
